onnxruntime/test/providers/qnn/matmul_test.cpp

```diff
@@ -108,268 +131,161 @@ static GetTestQDQModelFn<OutputQType> BuildQDQPerChannelMatMulTestCase(const Tes
 }
 
 //
 // CPU tests:
 //
-
-// TODO: Crashes during QNN CPU execution (QNN SDK 2.22)
-TEST_F(QnnCPUBackendTests, DISABLED_MatMulOp) {
-  RunMatMulOpOpTest(TestInputDef<float>({2, 3}, false, {-10.0f, -4.0f, -2.0f, 0.0f, 5.0f, 10.0f}),
-                    TestInputDef<float>({3, 2}, false, {-10.0f, -6.0f, -1.0f, 0.0f, 3.0f, 10.0f}),
-                    ExpectedEPNodeAssignment::All, 18);
-}
-
-// Test MatMul broadcasting
-// Failed randomly on Linux
-// Value of: expected_tensor.DataAsSpan<float>()
-// Expected: contains 896 values, where each value and its corresponding value in 16-byte object
-// <80-03 00-00 00-00 00-00 40-B8 53-08 CC-7F 00-00> are an almost-equal pair
-// Actual: 16-byte object <80-03 00-00 00-00 00-00 C0-B7 43-08 CC-7F 00-00>, where the value pair
-// (-5.19657087, 0) at index #29 don't match, which is 5.19657 from -5.19657
-TEST_F(QnnCPUBackendTests, DISABLED_MatMulOp_Broadcast) {
-  // Create two matrices with element values in the range [-10.0, 10.0].
-  std::vector<float> input_a = GetFloatDataInRange(-10.0f, 10.0f, 28 * 64);
-  std::vector<float> input_b = GetFloatDataInRange(-10.0f, 10.0f, 64 * 32);
-
-  RunMatMulOpOpTest(TestInputDef<float>({28, 1, 64}, false, input_a),
-                    TestInputDef<float>({64, 32}, false, input_b),
-                    ExpectedEPNodeAssignment::All, 18, 0.0004f);
-}
-
-// TODO: Crashes during QNN CPU execution (QNN SDK 2.22)
-TEST_F(QnnCPUBackendTests, DISABLED_MatMulOp_PaddingAndBroadcast_BLargerThanA) {
-  std::vector<int64_t> input0_shape = {2, 3, 2};
-  std::vector<int64_t> input1_shape = {3, 2, 2, 1};
-  RunMatMulOpOpTest(TestInputDef<float>(input0_shape, false, GetSequentialFloatData(input0_shape)),
-                    TestInputDef<float>(input1_shape, false, GetSequentialFloatData(input1_shape)),
-                    ExpectedEPNodeAssignment::All, 7);
+TEST_F(QnnCPUBackendTests, MatMulOp) {
+  RunMatMulOpTest(false, {2, 3}, {3, 2}, false, false);
+  RunMatMulOpTest(false, {2, 3}, {3, 2}, false, true);
+  RunMatMulOpTest(false, {2, 3}, {3, 2}, true, false);
+  RunMatMulOpTest(false, {2, 3}, {3, 2}, true, true);  // constant folding
+  RunMatMulOpTest(false, {2, 3}, {2, 3, 2}, false, false);
+  RunMatMulOpTest(false, {3, 3, 3}, {3, 2}, true, false);
+  RunMatMulOpTest(false, {2, 3, 3, 3}, {3, 2}, false, true);
+  RunMatMulOpTest(false, {2, 3, 3, 3}, {2, 3, 3, 2}, false, true);
+  RunMatMulOpTest(false, {2, 1, 2, 3}, {3, 3, 2}, false, false);
+  RunMatMulOpTest(false, {3}, {3}, false, false);
+  RunMatMulOpTest(false, {3}, {3}, false, true);
+  RunMatMulOpTest(false, {3}, {3}, true, false);
+  RunMatMulOpTest(false, {3}, {3, 2}, false, false);
+  RunMatMulOpTest(false, {3}, {3, 2}, false, true);
+  RunMatMulOpTest(false, {3}, {3, 3, 2}, true, false);
+  RunMatMulOpTest(false, {2, 3}, {3}, false, false);
+  RunMatMulOpTest(false, {2, 3}, {3}, true, false);
+  RunMatMulOpTest(false, {2, 3, 3, 3}, {3}, false, false);
+
+  // Failed randomly on Linux
```

Comment:
Is there a task to track this issue? If it affects a real model won't it invalidate the output which will be very hard to debug the root cause? 

is there an input combination we could special-case and not handle in the QNN EP until resolved? e.g. is it something like linux with input0 rank > 2 and input1 rank == 2? or would we need more testing to narrow down the problematic combination?

Reply:
It's platform, backend (CPU or HTP) and qnn version related. Actually most of the original UTs before the changes were disabled due to this error, some of them could pass using latest updated qnn version. I think it's issue from QNN code, but it's not possible for us to debug into it. Maybe we should report such problem to QNN.

Reply:
I think we should only be taking a node if we can guarantee the output will be correct. Trying to debug a production issue that was caused by this would be very difficult.

Reply:
We can do this if it's schema related. But if it's random data related, then end up we will not assign QNN EP to any MatMul, but MatMul is a very common and important Op for most of the models... And seems it's very hard and time consuming to get any conclusion on this.

Reply:
Would be great to ask the Qualcomm folks if they have any insight into the issue, especially if it might be fixed in a specific version of QNN.